src/orion/client/experiment.py

```diff
@@ -609,24 +642,70 @@ def workon(self, fct, max_trials=infinity, **kwargs):
 
         Raises
         ------
-        `ValueError`
-             If results returned by `fct` have invalid format
-        `orion.core.utils.exceptions.UnsupportedOperation`
+        :class:`orion.core.utils.exceptions.InvalidResult`
+             If results returned by `fct` have invalid format.
+
+        :class:`orion.core.utils.exceptions.WaitingForTrials`
+            if the experiment is not completed and algorithm needs to wait for some
+            trials to complete before it can suggest new trials.
+
+        :class:`orion.core.utils.exceptions.BrokenExperiment`
+            if too many trials failed to run and the experiment cannot continue.
+            This is determined by ``max_broken`` in the configuration of the experiment.
+
+        :class:`orion.core.utils.exceptions.SampleTimeout`
+            if the algorithm of the experiment could not sample new unique points.
+
+        :class:`orion.core.utils.exceptions.UnsupportedOperation`
             If the experiment was not loaded in executable mode.
 
         """
         self._check_if_executable()
 
+        if max_trials is None:
+            max_trials = orion.core.config.worker.max_trials
+
+        if max_broken is None:
+            max_broken = orion.core.config.worker.max_broken
+
+        worker_broken_trials = 0
         trials = 0
         kwargs = flatten(kwargs)
-        while not self.is_done and trials < max_trials:
-            trial = self.suggest()
-            if trial is None:
-                log.warning("Algorithm could not sample new points")
-                return trials
-            kwargs.update(flatten(trial.params))
-            results = fct(**unflatten(kwargs))
-            self.observe(trial, results=results)
+        while not self.is_done and trials - worker_broken_trials < max_trials:
```

Comment:
for single worker, `trials - worker_broken_trials < max_trials` seems always make `self.is_done` as `false` only if `worker_broken_trials ` > 0?

Reply:
I'm not sure I understand. Do you mean that if `trials - worker_broken_trials < max_trials` is true then automatically `self.is_done` cannot be true?

Here `max_trials` and `max_broken` can be different then those for the experiment.

Reply:
Forgot this,  I took it wrong, I saw `trials - worker_broken_trials` as `trials + worker_broken_trials`. :)

Reply:
haha ok :)